designs/nodepool-requirement-flexibility.md

```diff
@@ -0,0 +1,185 @@
+# NodePool Requirement Flexibility 
+
+**NodeClaim**: Karpenter’s representation of a Node request. A NodeClaim declaratively describes how the node should look, including instance type requirements that constrain the options that the Cloud Provider can launch with. For the AWS Cloud Provider, the instance requirements of the NodeClaim translate to the requirements of the CreateFleet request. More details can be found [here](https://github.com/aws/karpenter-provider-aws/blob/main/designs/node-ownership.md).
+
+**Preflight NodeClaim:** Karpenter’s in-memory representation of a NodeClaim that is computed during a provisioning loop. This is a NodeClaim that Karpenter hasn’t created yet CD, but is planning to launch at the end of the scheduling loop. This contains the full set of instance type possibilities which change (and become more constrained) as more pods are scheduled to the NodeClaim.
+
+## Motivation
+
+Karpenter’s scheduling algorithm uses a well-known bin-packing algorithm ([First Fit Descending](https://en.wikipedia.org/wiki/First-fit-decreasing_bin_packing)) to pack as many pods onto a set of instance type options as possible. This algorithm will necessarily continue packing pods onto Preflight Nodes so long as you have at least one instance type option that can fulfill all the pod requests.
+
+This methodology works fine for on-demand instance types, but starts to break down when you are requesting spot capacity. If you are launching spot capacity and want to ensure that you will not launch an instance that will immediately get interrupted after launch, you need to make sure that you have enough launch options in your launch request that your likelihood for launching a low-availability instance will be slim. Karpenter currently has no mechanism to ensure that today.
+
+This RFC proposes an additional key `minValues` in the NodePool requirements block, allowing Karpenter to be aware of user-specified flexibility minimums while scheduling pods to a cluster. If Karpenter cannot meet this minimum flexibility when scheduling a pod, it will fail the scheduling loop for that NodePool, either falling back to another NodePool which meets the pod requirements or failing scheduling the pod altogether.
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+spec:
+  template:
+    spec:
+      requirements:
+        - key: karpenter.kwok.sh.instance-family
+          operator: In
+          values: ["c", "m", "r"]
+          minValues: 2
+        - key: node.kubernetes.io/instance-type
+          operator: Exists
+          minValues: 10
+```
+
+## Use-Cases
+
+1. I want to configure Karpenter to enforce more flexibility when launching spot nodes, ensuring that I am constantly using enough instance types that I get a high-availability instance type
```

Comment:
Thinking about flexibility in launching nodes, or replacing nodes is there a place for minimum discount for consolidation? User may value the nodes just existing and not cycling up and down and paying a bit more, but might have a threshold where the cost makes the disruption worth it.

Reply:
>  there a place for minimum discount for consolidation

Yes! I think this is something that we were considering as part of the spot-to-spot consolidation design: https://github.com/kubernetes-sigs/karpenter/blob/main/designs/spot-consolidation.md?plain=1#L32. We ended up seeing this as orthogonal, but I still think that it's something that we want to do so that users can say I only want to be interrupted if it's going to give me _this_ much cost savings.